[RFC] Add an RFC for bit and byte conversions.

acoglio opened the pull request:
This is an attempt to collect and organize the information contained in various GitHub and Slack discussions into one coherent place for easy reference -- and update, if needed.

Please also review the last part of the 'Alternatives' section about naming. As always, when we add features, we should also keep in mind how they might fit within a broader consistent trajectory for the language.

gluax commented:
@acoglio that main way described in the RFC is horrific to implement. We would have to essentially to implement a [dummy call ](https://github.com/AleoHQ/leo/blob/7ae56f5c7c5b21223938a2d26c4533fbfa14e38d/asg/src/prelude.rs) and check for that dummy call in the [compiler ](https://github.com/AleoHQ/leo/blob/8e8e3fdf4215732b454193afa3d7d79f83c856b4/compiler/src/prelude/blake2s.rs) for each of those.

This is not only a nightmare way to implement something like that, but is silly for something planned for deprecation shortly.

I would suggest we can call the methods on variables with the types themselves only(meaning we can't call directly on types we would like to). As for from bytes we would just allow the user to call it on the array and has to have an expected type as well as the correct number of bytes and etc.

Though really we should maybe just stop everything and do types as circuits now before this, because either way leads to so much chaos that is detrimental to Leo in my opinion.

acoglio commented:
@gluax This could be a good discussion to have (when I'm back), on a call, and looking at the code in more detail.

I think I understand your concerns based on the links you sent, but I also think that it's possible to implement something like this succinctly and efficiently. The situation here of a family of 80 functions (= 10 types * 2 directions * 2 big-vs-little-endian * 2 bits-vs-bytes) with regularly structured names is somewhat similar to a situation in my ACL2-to-C compiler ([here](https://github.com/acl2/acl2/tree/master/books/kestrel/c/atc), if you are curious), where there are several such families, larger that this, one with 1,600 members (= 16 binary strict operations * 10 types for the left operand * 10 types for the right operand), and nowhere I wrote similar code replicated 1,600 times.

We might also need to take a step back and look how natively implemented Leo functions like Blake2s are handled currently, and whether there are better ways.

gluax commented:
As far as the implementation I discussed on 8/11/21's standup, please see #1248 as it covers the scenario.

acoglio commented:
@gluax I added an alternative in which we directly go with scalar type methods, which now seems the preferred approach given that we have a streamlined way to do methods on scalar types. If team consensus confirms that, we'll "swap" things in the RFC.

I also added a paragraph about implementing the current design in a concise manner, since as I mentioned i don't think that the current proposal necessarily implies a verbose implementation. We may well continue to agree to disagree on this particular point 😀  but it doesn't matter as we are likely going with methods directly, which sidesteps the whole issue.

acoglio commented:
I'm merging this in, so it's in master and we don't have to switch around branches. As always, this doesn't mean it's final, it just means it's in master.
